apollo-compiler/src/test/graphql/com/example/fragment_in_fragment/AllStarships.kt

```diff
@@ -86,57 +89,381 @@ class AllStarships : Query<AllStarships.Data, AllStarships.Data, Operation.Varia
+      )
 
-      operator fun invoke(reader: ResponseReader): Node = reader.run {
+      operator fun invoke(reader: ResponseReader): HomeworldImpl = reader.run {
         val __typename = readString(RESPONSE_FIELDS[0])!!
-        val fragments = Fragments(reader)
-        Node(
+        HomeworldImpl(
+          __typename = __typename
+        )
+      }
+
+      @Suppress("FunctionName")
+      fun Mapper(): ResponseFieldMapper<HomeworldImpl> = ResponseFieldMapper { invoke(it) }
+    }
+  }
+
+  /**
+   * A large mass, planet or planetoid in the Star Wars Universe, at the time of
+   * 0 ABY.
+   */
+  data class PlanetFragmentImpl(
+    /**
+     * The name of this planet.
+     */
+    override val name: String?,
+    override val __typename: String = "Planet"
+  ) : Homeworld, PlanetFragment, PilotFragment.Homeworld {
+    override fun marshaller(): ResponseFieldMarshaller {
+      return ResponseFieldMarshaller.invoke { writer ->
+        writer.writeString(RESPONSE_FIELDS[0], this@PlanetFragmentImpl.name)
+        writer.writeString(RESPONSE_FIELDS[1], this@PlanetFragmentImpl.__typename)
+      }
+    }
+
+    companion object {
+      private val RESPONSE_FIELDS: Array<ResponseField> = arrayOf(
+        ResponseField.forString("name", "name", null, true, null),
+        ResponseField.forString("__typename", "__typename", null, false, null)
+      )
+
+      operator fun invoke(reader: ResponseReader): PlanetFragmentImpl = reader.run {
+        val name = readString(RESPONSE_FIELDS[0])
+        val __typename = readString(RESPONSE_FIELDS[1])!!
+        PlanetFragmentImpl(
+          name = name,
+          __typename = __typename
+        )
+      }
+
+      @Suppress("FunctionName")
+      fun Mapper(): ResponseFieldMapper<PlanetFragmentImpl> = ResponseFieldMapper { invoke(it) }
+    }
+  }
+
+  /**
+   * A large mass, planet or planetoid in the Star Wars Universe, at the time of
+   * 0 ABY.
+   */
+  interface Homeworld : PilotFragment.Homeworld {
+    override val __typename: String
+
+    override fun marshaller(): ResponseFieldMarshaller
+
+    companion object {
+      private val RESPONSE_FIELDS: Array<ResponseField> = arrayOf(
+        ResponseField.forString("__typename", "__typename", null, false, null)
+      )
+
+      operator fun invoke(reader: ResponseReader): Homeworld {
+        val typename = reader.readString(RESPONSE_FIELDS[0])
+        return when(typename) {
+          "Planet" -> PlanetFragmentImpl(reader)
+          else -> HomeworldImpl(reader)
+        }
+      }
+    }
+  }
+
+  /**
+   * An individual person or character within the Star Wars universe.
+   */
+  data class PilotFragmentImpl(
+    /**
+     * The name of this person.
+     */
+    override val name: String?,
+    /**
+     * A planet that this person was born on or inhabits.
+     */
+    override val homeworld: Homeworld?,
+    override val __typename: String = "Person"
+  ) : Node1, PilotFragment, StarshipFragment.Node {
```

Comment:
I'm still trying to wrap my head around this but here inheriting from `StarshipFragment.Node` is not strictly required as it's already a super interface of `Node1`.

Reply:
Yeah makes sense such optimization